protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol

```diff
@@ -193,111 +311,96 @@ contract Weather is Sun {
-        uint256 instantaneousBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(instantaneousReserves, beanIndex, ratios, wellFunction.data);
-
-        uint256 currentBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(currentReserves, beanIndex, ratios, wellFunction.data);
+    // code review note: have some casual comments here for just understanding the code, happy to
+    // change this function to make it more efficient based on feedback.
+    function calculateSopPerWell(
+        WellDeltaB[] memory wellDeltaBs,
+        uint256 totalPositiveDeltaB,
+        uint256 totalNegativeDeltaB,
+        uint256 positiveDeltaBCount
+    ) public pure returns (WellDeltaB[] memory) {
+        // most likely case is that all deltaBs are positive
+        if (positiveDeltaBCount == wellDeltaBs.length) {
+            // if all deltaBs are positive, need to sop all to zero, so return existing deltaBs
+            return wellDeltaBs;
+        }
 
-        // Calculate the signed Sop beans for the two reserves.
-        int256 lowestSopBeans = int256(instantaneousBeansAtPeg).sub(
-            int256(instantaneousReserves[beanIndex])
-        );
-        int256 currentSopBeans = int256(currentBeansAtPeg).sub(int256(currentReserves[beanIndex]));
+        if (positiveDeltaBCount == 0) {
+            // No positive values, so no well flooding needed, return zeros
+            for (uint256 i = 0; i < positiveDeltaBCount; i++) {
+                wellDeltaBs[i].deltaB = 0;
+            }
+            return wellDeltaBs;
+        }
 
-        // Use the minimum of the two.
-        if (lowestSopBeans > currentSopBeans) {
-            lowestSopBeans = currentSopBeans;
+        if (totalPositiveDeltaB < totalNegativeDeltaB) {
+            // This can occur if the twaDeltaB is positive, but the instanteous deltaB is negative or 0
+            // In this case, no reductions are needed.
+            for (uint256 i = 0; i < positiveDeltaBCount; i++) {
+                wellDeltaBs[i].deltaB = 0;
+            }
+            return wellDeltaBs;
         }
 
-        // If the sopBeans is negative, the sop should not occur.
-        if (lowestSopBeans < 0) return (0, IERC20(address(0)));
+        uint256 shaveToLevel = totalNegativeDeltaB / positiveDeltaBCount;
 
-        // SafeCast not necessary due to above check.
-        sopBeans = uint256(lowestSopBeans);
+        for (uint256 i = positiveDeltaBCount; i > 0; i--) {
+            if (shaveToLevel > uint256(wellDeltaBs[i - 1].deltaB)) {
+                shaveToLevel += (shaveToLevel - uint256(wellDeltaBs[i - 1].deltaB)) / (i - 1);
 
-        // the sopToken is the non bean token in the well.
-        sopToken = tokens[beanIndex == 0 ? 1 : 0];
+                // reduction amount does not need to be set here,
+                // but is written for demonstration.
+                wellDeltaBs[i - 1].deltaB = 0;
```

Comment:
remove here? or lets comment

Reply:
Updated/commented here: https://github.com/BeanstalkFarms/Beanstalk/pull/870/commits/ed3adec05bf0cacc4641a92b66fa81aeafb0ea06

Reply:
This actually ended up breaking the test, now that deltaB is repurposed this variable does need updating to zero, otherwise logic breaks. Fixed here: https://github.com/BeanstalkFarms/Beanstalk/pull/870/commits/1d2d4f3a683125f91baa0202f66af05911114d0f